tiatoolbox/tools/patchextraction.py

```diff
@@ -0,0 +1,268 @@
+
+        Args:
+            patches: patch-level predictions
+
+        Returns:
+            image: merged prediction
+
+        """
+        raise NotImplementedError
+
+
+class FixedWindowPatchExtractor(PatchExtractor):
+    """Extract and merge patches using fixed sized windows for images and labels.
+
+    Args:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+
+    Attributes:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+
+    """
+
+    def __init__(
+        self,
+        input_img,
+        patch_size,
+        resolution=0,
+        units="level",
+        stride=1,
+    ):
+        super().__init__(
+            input_img=input_img,
+            patch_size=patch_size,
+            resolution=resolution,
+            units=units,
+        )
+        self.stride = stride
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def merge_patches(self, patches):
+        raise NotImplementedError
+
+
+class VariableWindowPatchExtractor(PatchExtractor):
+    """Extract and merge patches using variable sized windows for images and labels.
+
+    Args:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+        label_patch_size(Tuple of int): network output label (width, height).
+
+    Attributes:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+        label_patch_size(Tuple of int): network output label (width, height).
+
+    """
+
+    def __init__(
+        self,
+        input_img,
+        patch_size,
+        resolution=0,
+        units="level",
+        stride=1,
+        label_patch_size=None,
+    ):
+        super().__init__(
+            input_img=input_img,
+            patch_size=patch_size,
+            resolution=resolution,
+            units=units,
+        )
+        self.stride_h = stride
+        self.label_patch_size = label_patch_size
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def merge_patches(self, patches):
+        raise NotImplementedError
+
+
+class PointsPatchExtractor(PatchExtractor):
+    """Extracting patches with specified points as a centre.
+
+    Args:
+        locations_list(ndarray, pd.DataFrame, str, pathlib.Path): contains location
+         and/or type of patch. Input can be path to a csv or json file.
+        num_examples_per_patch(int): Number of examples per patch for ensemble
+         classification, default=9 (centre of patch and all the eight neighbours as
+         centre).
+
+    Attributes:
+        locations_list(pd.DataFrame): A table containing location and/or type of patch.
+        num_examples_per_patch(int): Number of examples per patch for ensemble
+         classification.
+
+    """
+
+    def __init__(
+        self,
+        input_img,
+        locations_list,
+        patch_size=224,
+        resolution=0,
+        units="level",
+        num_examples_per_patch=9,
```

Comment:
I suggest the default value for `num_examples_per_patch` be 1, as model ensembling should be optional.

Reply:
Agreed, this was to allow for augmentation as in SCCNN paper. But current implementation allows only 1 example.

Reply:
Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/commit/8e3bbe7dc1e1807dd0b7fc4bd558ddd1bd56c585